apps/c_gpio/main.c

```diff
@@ -0,0 +1,78 @@
+/**
+ * This application is for testing GPIO interrupts in the nRF51822 EK.
+ * To run this application, hook up a button connected to VDD to GPIO pin 1
+ * (the top right pin on the top left header).
+ *  
+ * When it boots, you should see one of the two LEDs blink 5 times, then
+ * go silent. This is to show that the app has booted correctly.
+ *  
+ * Then, when you push the button, the other LED should blink.
+ *
+ * Note that to an application, pin 0 is LED0 (GPIO pin 18), pin 1 is 
+ * LED1 (GPIO pin 19), and pin 3 is GPIO pin 0.
+ */
+
+#include <firestorm.h>
+#include <gpio.h>
+
+#define LED_1 1
+
+/* Delay for for the given microseconds (approximately).
+ *
+ * For a 16 MHz CPU, 1us == 16 instructions (assuming each instruction takes
+ * one cycle). */
+static void busy_delay_us(int duration)
+{
+	// The inner loop instructions are: 14 NOPs + 1 SUBS/ADDS + 1 CMP
+	while (duration-- != 0) {
+		__asm volatile (
+			"nop\n"
+			"nop\n"
+			"nop\n"
+			"nop\n"
+			"nop\n"
+			"nop\n"
+			"nop\n"
+			"nop\n"
+			"nop\n"
+			"nop\n"
+			"nop\n"
+			"nop\n"
+			"nop\n"
+			"nop\n"
+		);
+	}
+}
+
+/* Delay for for the given milliseconds (approximately).
+ *
+ * Note that this is not precise as there are 2 extra instructions on the inner
+ * loop. Therefore, there is 1us added every 8 iterations. */
+static void busy_delay_ms(int duration) {
+	while (duration-- != 0) {
+		busy_delay_us(1000);
+                yield();
```

Comment:
Neither. It will push an event handler _only_ on a call to `wait`.

Previously (until changes to master in the last week), the kernel itself would also be blocked until a system call so it would have totally missed a GPIO interrupt in this case (yield would have helped there). But that's fixed now.